### DISCUSSION TITLE
Class/subclass issue when adding GeoJSON to tilemaker.

### DISCUSSION THREAD

**Discussion by hamadm123** [systemed/tilemaker#646] [Q&A] [answered]
Hello - I am adding a geoJson of subway lines my maptiles of NYC. When I render the map, I see the geoJSON addition fine, however, the class is somehow "ocean." I have a feeling the answer may be a bit obvious - but I am having a hard time figuring it out, any help would be appreciated

<img width="851" alt="Screenshot 2024-01-17 at 10 47 44 AM" src="https://github.com/systemed/tilemaker/assets/20972267/1a5bf93e-e741-40f5-b735-de4afd263ab4">

Please see below for the config file:

	"layers": {
		"place":            { "minzoom":  0, "maxzoom": 14 },
		"boundary":         { "minzoom":  0, "maxzoom": 14, "simplify_below": 12, "simplify_level": 0.0003, "simplify_ratio": 2 },

		"poi":              { "minzoom": 12, "maxzoom": 14 },
		"poi_detail":       { "minzoom": 14, "maxzoom": 14, "write_to": "poi"},

		"housenumber":      { "minzoom": 14, "maxzoom": 14 },

		"waterway":         { "minzoom":  8,  "maxzoom": 14, "simplify_below": 12, "simplify_level": 0.0003, "simplify_ratio": 2 },
		"waterway_detail":  { "minzoom": 12,  "maxzoom": 14, "write_to": "waterway" },

		"transportation":             { "minzoom": 4,  "maxzoom": 14, "simplify_below": 13, "simplify_level": 0.0003 },
		"transportation_name":        { "minzoom": 8,  "maxzoom": 14 },
		"transportation_name_mid":    { "minzoom": 12, "maxzoom": 14, "write_to": "transportation_name" },
		"transportation_name_detail": { "minzoom": 14, "maxzoom": 14, "write_to": "transportation_name" },

		"building":          { "minzoom": 13, "maxzoom": 14 },

		"water":             { "minzoom": 6,  "maxzoom": 14, "simplify_below": 12, "simplify_level": 0.0003, "simplify_ratio": 2},
		"ocean":             { "minzoom": 0,  "maxzoom": 14, "source": "coastline/water_polygons.shp", "filter_below": 12, "filter_area": 0.5, "simplify_below": 13, "simplify_level": 0.0001, "simplify_ratio": 2, "write_to": "water" },
		"water_name":        { "minzoom": 14, "maxzoom": 14 },
		"water_name_detail": { "minzoom": 14, "maxzoom": 14, "write_to": "water_name" },

		"aeroway":           { "minzoom": 11, "maxzoom": 14 },
		"aerodrome_label":   { "minzoom": 10, "maxzoom": 14 },
		"park":              { "minzoom": 11, "maxzoom": 14 },
		"landuse":           { "minzoom":  4, "maxzoom": 14, "simplify_below": 13, "simplify_level": 0.0003, "simplify_ratio": 2 },
		"urban_areas":       { "minzoom":  4, "maxzoom":  8, "source": "landcover/ne_10m_urban_areas/ne_10m_urban_areas.shp", "source_columns": ["featurecla"], "simplify_below": 7, "simplify_level": 0.0003, "simplify_ratio": 2, "write_to": "landuse" },
		"landcover":         { "minzoom":  0, "maxzoom": 14, "simplify_below": 13, "simplify_level": 0.0003, "simplify_ratio": 2 },
		"ice_shelf":         { "minzoom":  0, "maxzoom":  9, "source": "landcover/ne_10m_antarctic_ice_shelves_polys/ne_10m_antarctic_ice_shelves_polys.shp", "source_columns": ["featurecla"], "simplify_below": 13, "simplify_level": 0.0005, "write_to": "landcover" },
		"glacier":           { "minzoom":  2, "maxzoom":  9, "source": "landcover/ne_10m_glaciated_areas/ne_10m_glaciated_areas.shp", "source_columns": ["featurecla"], "simplify_below": 13, "simplify_level": 0.0005, "write_to": "landcover" },
		"mountain_peak":     { "minzoom": 11, "maxzoom": 14 },
		"subways":           { "minzoom": 11, "maxzoom": 14, "source": "geoJson/SubwayLines.geojson", "source_columns": ["name"]}
	},
	"settings": {
		"minzoom": 0,
		"maxzoom": 14,
		"basezoom": 14,
		"include_ids": false,
		"combine_below": 14,
		"name": "Tilemaker to OpenMapTiles schema",
		"version": "3.0",
		"description": "Tile config based on OpenMapTiles schema",
		"compress": "gzip",
		"filemetadata": {
			"tilejson": "2.0.0", 
			"scheme": "xyz", 
			"type": "baselayer", 
			"format": "pbf", 
            "tiles": ["https://example.com/liechtenstein/{z}/{x}/{y}.pbf"]
		}
	}

**Comment by cldellow** [comment, marked as answer]
You probably want to tweak your `attribute_function`. It gets called when processing GeoJSON/shapefiles - you receive the layer name and the attributes, and decide which attributes you want to write out.

The openmaptiles profile defaults to `ocean`, so I'm guessing you've maybe started from that file: https://github.com/systemed/tilemaker/blob/172d13327cfa91dbdf0107d6110e1924be3280aa/resources/process-openmaptiles.lua#L654-L665

**Reply by hamadm123** [reply]
Thanks - that was the right place. If anyone else is struggling with these and comes across this I simply printed out the inputs to write a new elseif, see below:

```
  function attribute_function(attr, layer)
      -- Print the layer
      print("Layer:", layer)
  
      -- Print all key-value pairs in attr
      for key, value in pairs(attr) do
          print("attr[" .. key .. "] =", value)
      end
  
      if attr["featurecla"] == "Glaciated areas" then
          return { subclass = "glacier" }
      elseif attr["featurecla"] == "Antarctic Ice Shelf" then
          return { subclass = "ice_shelf" }
      elseif attr["featurecla"] == "Urban area" then
          return { class = "residential" }
      else
          return { class = "ocean" }
      end
  end
```

**Comment by systemed** [comment]
I've tweaked this in https://github.com/systemed/tilemaker/commit/c490b5c744f9efba3b53200f94253713645a352f so that you can just add GeoJSON files to the default config, and it won't assume they're the ocean.
